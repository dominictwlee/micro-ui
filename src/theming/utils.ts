const INVALID_CSS_ERROR = new Error('Invalid CSS value: Expects a string');

export function toCssVarNames<T>(themeObj: T, prefix = '-'): T {
  const vars: any = {};
  for (const [key, value] of Object.entries(themeObj)) {
    const varName = `${prefix}-${key}`;
    if (typeof value === 'object') {
      vars[key] = toCssVarNames(value, varName);
    } else if (typeof value === 'string') {
      vars[key] = `var(${varName})`;
    } else {
      throw INVALID_CSS_ERROR;
    }
  }
  return vars;
}

interface CssVarMap {
  [cssVarName: string]: string;
}

export function toCssVars<T>(themeObj: T, prefix = '-') {
  const vars: CssVarMap = {};

  for (const [key, value] of Object.entries(themeObj)) {
    const varName = `${prefix}-${key}`;
    if (typeof value === 'object') {
      const nestedVars = toCssVars(value, varName);
      for (const [nestedKey, nestedValue] of Object.entries(nestedVars)) {
        vars[nestedKey] = nestedValue;
      }
    } else if (typeof value === 'string') {
      vars[varName] = value;
    } else {
      throw INVALID_CSS_ERROR;
    }
  }
  return vars;
}
