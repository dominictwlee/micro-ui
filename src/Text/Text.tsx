import React, { HTMLProps, PropsWithChildren } from 'react';
import { css } from '@emotion/react';

import { themeVars } from '../theming';

const elementTags = {
  h1: 'h1',
  h2: 'h2',
  h3: 'h3',
  h4: 'h4',
  h5: 'h5',
  h6: 'h6',
  h7: 'h6',
  p1: 'p',
} as const;

export interface TextProps
  extends HTMLProps<HTMLParagraphElement | HTMLHeadingElement> {
  variant?: keyof typeof elementTags;
}

const base = ({ variant = 'p1' }: TextProps) => css`
  font-size: ${themeVars.typography[variant].fontSize};
  font-weight: ${themeVars.typography[variant].fontWeight};
  line-height: ${themeVars.typography[variant].lineHeight};
`;

export default function Text({
  variant = 'p1',
  ...props
}: PropsWithChildren<TextProps>) {
  const Element = elementTags[variant];
  return <Element {...props} css={base} />;
}
