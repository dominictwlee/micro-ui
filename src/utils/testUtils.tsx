import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from '../theming';

const Providers = ({ children }: any) => {
  return <ThemeProvider theme="light">{children}</ThemeProvider>;
};

const customRender = (ui: any, options: any = {}) =>
  render(ui, { wrapper: Providers, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
