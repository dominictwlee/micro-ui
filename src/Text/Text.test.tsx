import React from 'react';

import Text from './Text';
import { render, screen } from '../utils/testUtils';

function Variants() {
  return (
    <div>
      {['h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'h7', 'p1'].map(el => (
        <Text key={el} variant={el as any}>
          {el}
        </Text>
      ))}
    </div>
  );
}

describe('Text', () => {
  test('render with children', () => {
    render(<Text>hello world</Text>);
    expect(screen.getByText('hello world')).toBeInTheDocument();
  });

  test('render different variants', () => {
    const { container } = render(<Variants />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
