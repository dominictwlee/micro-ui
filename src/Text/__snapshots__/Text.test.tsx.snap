// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text render different variants 1`] = `
<div>
  <h1
    class="css-1o7bj83-base"
  >
    h1
  </h1>
  <h2
    class="css-4icqhe-base"
  >
    h2
  </h2>
  <h3
    class="css-uoe7du-base"
  >
    h3
  </h3>
  <h4
    class="css-1llivza-base"
  >
    h4
  </h4>
  <h5
    class="css-1dzsr6v-base"
  >
    h5
  </h5>
  <h6
    class="css-1frn9l6-base"
  >
    h6
  </h6>
  <h6
    class="css-122qqrq-base"
  >
    h7
  </h6>
  <p
    class="css-1oykwkb-base"
  >
    p1
  </p>
</div>
`;
